### ISSUE TITLE
collapsible_match suggestion changes behaviour

### ISSUE THREAD

**Issue by glts** [rust-lang/rust-clippy#6740]
Lint name: collapsible_match

Clippy suggests collapsing the nested match in the following code snippet (the full code is [here](https://gitlab.com/glts/spf-milter/-/blob/8cd23bb27df8316edb15bd7e3a3f19d7596044b6/src/config/read.rs#L291-299)):

```rust
pub fn focus_error(error: &ConfigError) -> &dyn Error {
    match error {
        ConfigError::ReadConfig(e) => match e {
            ReadConfigError::Parse(e) => e,
            e => e,
        },
        e => e,
    }
}
```

I expected to see this happen: no Clippy warning

Instead, this happened:

```
warning: Unnecessary nested match
   --> src/config/read.rs:293:39
    |
293 |           ConfigError::ReadConfig(e) => match e {
    |  _______________________________________^
294 | |             ReadConfigError::Parse(e) => e,
295 | |             e => e,
296 | |         },
    | |_________^
    |
    = note: `#[warn(clippy::collapsible_match)]` on by default
help: The outer pattern can be modified to include the inner pattern.
   --> src/config/read.rs:293:33
    |
293 |         ConfigError::ReadConfig(e) => match e {
    |                                 ^ Replace this binding
294 |             ReadConfigError::Parse(e) => e,
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^ with this pattern
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_match
```

This suggestion is invalid. It changes the type of the error object is returned and noticeably changes the behaviour of the code (in this case the error message printed to the command-line changes).

### Meta

- `cargo clippy -V`: clippy 0.0.212 (cb75ad5 2021-02-10)
- `rustc -Vv`:
  ```
  rustc 1.50.0 (cb75ad5db 2021-02-10)
  binary: rustc
  commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
  commit-date: 2021-02-10
  host: x86_64-unknown-linux-gnu
  release: 1.50.0
  ```
